net/xia/ppal_ether/main.c

```diff
@@ -448,6 +455,69 @@ static const xia_ppal_all_rt_eops_t ether_all_rt_eops = {
 	},
 };
 
+/* Creates a cached hdr and alocates to a mfxid. */
+static int xia_ether_header_cache(const struct fib_xid_ether_main *mfxid,
+				  struct hh_cache *hh, __be16 type)
+{
+	struct ethhdr *eth;
+	const struct net_device *dev = mfxid->host_interface;
+
+	eth = (struct ethhdr *)
+	    (((u8 *)hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+
+	if (type != htons(ETH_P_XIP))
+		return -1;
+
+	eth->h_proto = type;
+	memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+	memcpy(eth->h_dest, mfxid->ha, ETH_ALEN);
+	hh->hh_len = ETH_HLEN;
+	return 0;
+}
+
+/* Updates the cached hdr on change of address
+* type : 1 -> destionation addr
+*        0 -> source addr
+*/
+static void xia_ether_header_cache_update(struct hh_cache *hh,
+					  const struct net_device *dev,
+					  const unsigned char *haddr,
+					  const int type)
+{
+	struct ethhdr *eth;
+
+	eth = (struct ethhdr *)
+		    (((u8 *)hh->hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+	if (type)
+		memcpy(eth->h_dest, haddr, ETH_ALEN);
+	else
+		memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+}
+
+/* Cached Hdr functions. */
+static int mfxid_hh_init(struct fib_xid_ether_main *mfxid)
+{
+	__be16 prot = htons(ETH_P_XIP);
+	struct hh_cache	*hh = &mfxid->cached_hdr;
+
+	if (!hh->hh_len)
+		return xia_ether_header_cache(mfxid, hh, prot);
+	return 1;
+}
+
+static void mfxid_update_hhs(struct fib_xid_ether_main *mfxid, const int type)
+{
+	struct hh_cache *hh;
+
+	hh = &mfxid->cached_hdr;
```

Comment:
You forgot to add field `cached_hdr` in `struct fib_xid_ether_main`.

This concerns me because it means that you haven't bothered to compile the code after all the changes you made, let alone test the code before asking for a code review.

Reply:
I'm really sorry Michel. I compiled the code before the last step of shifting the data from header file to c file, which i'm doing currently and got to know about a few minutes earlier. The problem lies with the cached header struct only in which somehow the cached_hdr didn't get included while copy/pasting(don't know why), the rest of the code compiles really well. I will be sending screenshots too just in about an hour.

Reply:
Okay. I'm going to review the newer code. Thank you for the explanation.